src/ast/passes/semantic_analyser.h

```diff
@@ -105,11 +105,47 @@ class SemanticAnalyser : public Visitor {
   int analyse();
 
 private:
+  class PassTracker {
```

Comment:
nit: Move out of line of SemanticAnalyser definition? Just to keep the class stuff more compact